File: src/main/java/io/zeitwert/ddd/aggregate/model/base/AggregateRepositorySPI.java
package io.zeitwert.ddd.aggregate.model.base;

import io.crnk.core.queryspec.QuerySpec;
import io.zeitwert.ddd.aggregate.model.Aggregate;
import io.zeitwert.ddd.aggregate.model.AggregatePersistenceProvider;
import io.zeitwert.ddd.property.model.PropertyProvider;

import java.util.List;

import org.jooq.Record;

/**
 * This class defines the internal callbacks for a AggregateRepository
 * implementation.
 */
public interface AggregateRepositorySPI<A extends Aggregate, V extends Record> {

	/**
	 * Get the PropertyProvider for this repository
	 *
	 * @return PropertyProvider
	 */
	PropertyProvider getPropertyProvider();

	/**
	 * Get the PersistenceProvider for this repository
	 *
	 * @return AggregatePersistenceProvider
	 */
	AggregatePersistenceProvider<A> getPersistenceProvider();

	/**
	 * Create a new Aggregate instance (technical instantiation)
	 *
	 * @param aggregateState aggregate state
	 * @return new aggregate
	 */
	A newAggregate(Object aggregateState);

	/**
	 * Register the required PartRepositories
	 */
	void registerPartRepositories();

	/**
	 * Initialize Parts of new Aggregate instance
	 *
	 * @param aggregate aggregate
	 */
	void doInitParts(A aggregate);

	/**
	 * Do some work after create, f.ex. fire events
	 *
	 * @param aggregate aggregate
	 */
	void doAfterCreate(A aggregate);

	/**
	 * Load Parts from database and attach to Aggregate
	 *
	 * @param aggregate the aggregate
	 */
	void doLoadParts(A aggregate);

	/**
	 * Do some work after load, f.ex. fire events
	 *
	 * @param aggregate aggregate
	 */
	void doAfterLoad(A aggregate);

	/**
	 * Do some work before store, f.ex. make sure certain structures are in place
	 *
	 * @param aggregate aggregate
	 */
	void doBeforeStore(A aggregate);

	/**
	 * Store Parts
	 *
	 * @param aggregate aggregate
	 */
	void doStoreParts(A aggregate);

	/**
	 * Do some work after store, f.ex. fire events
	 *
	 * @param aggregate aggregate
	 */
	void doAfterStore(A aggregate);

	/**
	 * Find aggregates (ids) according to given Query
	 *
	 * @param querySpec query
	 */
	List<V> doFind(QuerySpec querySpec);

}

File: fm-dddrive/src/main/java/dddrive/ddd/core/model/AggregateRepositorySPI.kt
package dddrive.ddd.core.model

/**
 * This class defines the internal callbacks for an AggregateRepository
 * implementation.
 */
interface AggregateRepositorySPI<A : Aggregate> {

	/**
	 * Register the parts of the aggregate
	 */
	fun registerParts()

	/**
	 * Get the PersistenceProvider for this repository
	 *
	 * @return AggregatePersistenceProvider
	 */
	val persistenceProvider: AggregatePersistenceProvider<A>

	/**
	 * Create a new aggregate instance. Concrete repositories must override this to directly
	 * instantiate their Impl class.
	 */
	fun createAggregate(isNew: Boolean): A

	/**
	 * Do some work after create, f.ex. fire events
	 *
	 * @param aggregate aggregate
	 */
	fun doAfterCreate(aggregate: A)

	/**
	 * Do some work after load, f.ex. fire events
	 *
	 * @param aggregate aggregate
	 */
	fun doAfterLoad(aggregate: A)

	/**
	 * Do some work before store, f.ex. make sure certain structures are in place
	 *
	 * @param aggregate aggregate
	 */
	fun doBeforeStore(aggregate: A)

	/**
	 * Do some work after store, f.ex. fire events
	 *
	 * @param aggregate aggregate
	 */
	fun doAfterStore(aggregate: A)

}
